teps/0042-taskrun-breakpoint-on-failure.md

```diff
@@ -0,0 +1,102 @@
+- [Design Details](#design-details)
+- [Test Plan](#test-plan)
+- [Design Evaluation](#design-evaluation)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Infrastructure Needed (optional)](#infrastructure-needed-optional)
+- [Upgrade &amp; Migration Strategy (optional)](#upgrade--migration-strategy-optional)
+- [References (optional)](#references-optional)
+<!-- /toc -->
+
+## Summary
+
+If a TaskRun fails in Tekton, it can be hard to understand at what step and what exactly in the step led to the failure.
+To solve this problem, it can be possible for the users to set a breakpoint on failure specification in the TaskRun which
+halts it at a particular step, making it easy for the user to debug it. The following proposal covers the implementation
+details regarding the same.
+
+## Motivation
+
+With the addition of a "breakpoint on failure" mechanism for TaskRuns, the user experience should vastly improve as traditional
+CI/CD systems have only a primitive debugging present, whereas while leveraging containers we can make Live Debugging in
+Tekton possible with this.
+
+The user would enjoy 
+
+- Improved Experience
+- Faster iteration during Pipeline development
+- Real time debugging 
+
+### Goals
+
+When the "breakpoint on failure" spec is mentioned in a particular TaskRun it should
+- Pause TaskRun on failure of a step
+- Allow User to get shell access to a debugging environment
+- Manually continue the TaskRun from where the breakpoint occurred
+
+### Non-Goals
+
+- PipelineRun debug/breakpoint
+- Breakpoint on other steps apart from the failing one (not in scope)
+
+## Requirements
+
+Understanding of Taskrun Controller and Entrypoint mechanism.
+
+## Proposal 
+
+To enable breakpoints in TaskRun, the Pod created by the TaskRun Controller can be augmented to add a `breakpoint` container in the Pod from where the user can debug the TaskRun.
+
+The implementation can be broken up into multiple stages.
+
+### Stage 1: Breakpoint onFailure
+
+To breakpoint and pause a TaskRun upon the failure of a Step, the TaskRun Pod which is created, would contain a `breakpoint` container which would run till the the TaskRun succeeds so that in case there is a failure, the next step would not start and the TaskRun would be halted.
+
+This can be achieved with the following :
+- Skip entrypoint step skipping if previous step has failed (Add -breakpoint_on_failure to entrypointer)
+- Add continue script for breakpoint container which the user can use to mark the failed step as successful
+- Mount all other container volumes in /tekton/debug/steps to check and debug information in other containers 
+
+Breakpoint Spec
+```yaml
+breakpoint:
+  onFailure: true
+```
+
+### Stage 2: Breakpoint onStep
+
+To breakpoint before a particular step we would mention `onStep` which would allow the user to halt the step before executing the breakpoint. 
+
+Breakpoint Spec
+```yaml
+breakpoint:
+  onStep:
+  - step1
+  - step4
+```
+
```

Comment:
Helpful addition on the proposal section @waveywaves 

I suspect there is more you can add around Design Details / Test Plan / Drawbacks / Alternatives

Per https://github.com/tektoncd/community/blob/main/teps/tools/tep-template.md.template

WDYT?

Reply:
Just learned the process has changed :-)

That process has been documented, but an update to the template may be coming near term

/lgtm